staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go

```diff
@@ -92,6 +92,10 @@ type YAMLDecoder struct {
 // the caller in framing the chunk.
 func NewDocumentDecoder(r io.ReadCloser) io.ReadCloser {
 	scanner := bufio.NewScanner(r)
+	// Adding maxCapacity to set the Buffer size to 5M to fix #89016
+	maxCapacity := 5120 * 1024
+	buf := make([]byte, maxCapacity)
```

Comment:
this allocates 5MB of memory on every yaml document read? That seems less than optimal. 

Is there an alternative method that could stream/split without preallocating large chunks like this?

Reply:
From the scanner docs:

> Programs that need more control over error handling or large tokens, or must run sequential scans on a reader, should use bufio.Reader instead.

That would be more what I expected

Reply:
alternatively, it looks like you can pass a smaller buffer and a larger max size, and the scanner will grow the buffer up to the max size if required. That would be an equally minimal fix and more acceptable. 

This also needs an accompanying test demonstrating the issue and fix.

Reply:
```
scanner := bufio.NewScanner(r)
	buf := make([]byte, 64*1024)
	scanner.Buffer(buf, 5*1024*1024)
	scanner.Split(splitYAMLDocument)
	return &YAMLDecoder{
		r:       r,
		scanner: scanner,
	}
```
how about it?How big is the buffer set? What about the maximum?

Reply:
> this allocates 5MB of memory on every yaml document read? 

No I don't think this is what happens. So from the [Buffer SourceCode](https://golang.org/src/bufio/scan.go?s=9061:9106#L252)  it assigns the initial s.buf to cap(buf)
```
s.buf = buf[0:cap(buf)]
s.maxTokenSize = max
```
and then on Scan() It starts from s.buf and builds a buffer size up to maxTokenSize [SourceCode](https://github.com/golang/go/blob/master/src/bufio/scan.go#L188)
```
newSize := len(s.buf) * 2
if newSize == 0 {
    newSize = startBufSize
}
if newSize > s.maxTokenSize {
    newSize = s.maxTokenSize
}
```

So from my understanding, it's not assigning 5MB of memory to every yaml. it assigns it Gradually for the large documents that need it. Please correct me if I misunderstood something.

Reply:
```
buf := make([]byte, maxCapacity)
``` 
This line allocates 5MB of memory, regardless of what scanner does

Reply:
Oh I missed that. Changed it now. like @tanjunchen suggested